doc/_build/man/bioplot.1
.\" Man page generated from reStructuredText.
.
.TH "BIOPLOT" "1" "May 30, 2017" "1.2.1" "bioplot"
.SH NAME
bioplot \- bioplot Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
Contents:
.SH INTRODUCTION TO BIOPLOT
.sp
You will not be able to use this program or any of its parts to slay a dragon, that I\(aqm sure of, and I do not guarantee
that it is fit for any other purpose at all, but I would appreciate if you would include a reference to this code and my name
in any publications you may write using its features or add a source reference to your code if you include some part(s)
of this code in yours. Especially if you plan to use part of the code please respect the rst_license\&.
.sp
bioplot.py is a program which can draw several plots that can be used
when evaluating the performance of a biometric system. It reads settings from the file \(aqbioplot.cfg\(aq.
But using the \-c option you can specify your own settings file (see rst_settings).
These settings may be used to determine the way information is shown in a plot, what directory plot files
are written to etc.
.sp
The plot types currently supported are: rst_accuracy, rst_detplot, rst_eerplot, rst_histogram,
rst_matrixplot, rst_rankingplot, rst_rocplot, rst_tippetplot and rst_zooplot\&.
.sp
Please read INSTALL.txt, this html documentation (note that the file \(aqmanual.txt\(aq is obsolete) and bioplot.cfg before you try to use
bioplot.py. You\(aqll learn more of the program\(aqs potential than from its command line help message.
.sp
@Windows dudes and dudettes: I\(aqm afraid you have to run the program by hand from a command line or build shortcuts
which not only start the program but also provide the command line options and parameters needed. You\(aqre on your
own here. I\(aqm a linux/OSX command line junky anyway so I did not spend any time building a (windows) gui. But don\(aqt fret. All plots
are shown in an interactive window, you can click on that as much as you like. The interface allows you to zoom in,
click on a point to see the associated label, save the plot in a file etc. etc.
.SS Usage
.sp
What bioplot.py basically does is: read a data file and plot one or more interactive graphs.
You can run bioplot.py from the command line. The follow command will show help info for bioplot:
On OSX and linux run:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-h
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
On Ms Windows run:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python.exe bioplot.py \-h
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This wil output the following information:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: bioplot.py [\-h] [\-Z] [\-A] [\-D] [\-E] [\-T] [\-M] [\-O] [\-R] [\-C] [\-H] [\-k]
                  [\-L LABELS [LABELS ...]] [\-e EXPNAME]
                  [\-i FILENAMES [FILENAMES ...]] [\-t DATATYPE] [\-d THRESHOLD]
                  [\-c CONFIGFILENAME] [\-l] [\-s] [\-q]

bioplot.py [plot type] [<label1> <label2> <label3> ...] bioplot.py version
This is bioplot.py version 1.2.1, Copyright (C) 2014: Jos Bouten
This program comes with ABSOLUTELY NO WARRANTY; for details run
\(gabioplot.py \-l\(aq. This is free software, and you are welcome to redistribute it
under certain conditions; type \(gabioplot.py \-l\(aq for details. This program was
written by Jos Bouten. You can contact me via josbouten at gmail dot com.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-Z, \-\-zoo             show zoo plot
  \-A, \-\-accuracy        show accuracy plot
  \-D, \-\-det             show Det plot
  \-E, \-\-eer             show EER plot
  \-T, \-\-tippet          show Tippett plot
  \-M, \-\-matrix          show matrix plot
  \-O, \-\-roc             show roc plot
  \-R, \-\-ranking         show ranking plot
  \-C, \-\-histogramc      show cumulative histogram
  \-H, \-\-histogram       show histogram
  \-k, \-\-kernel          show kernel estimate in histogram
  \-L LABELS [LABELS ...], \-\-label LABELS [LABELS ...]
                        add labels to plot
  \-e EXPNAME, \-\-exp EXPNAME
                        name of experiment used in plot title, default = test
  \-i FILENAMES [FILENAMES ...], \-\-inputfile FILENAMES [FILENAMES ...]
                        filename of filenames of data file(s) or name of
                        database, default = input/testdata_A.txt
  \-t DATATYPE, \-\-type DATATYPE
                        type of data, default = type3, use \(aqdatabase\(aq if you
                        want to read data from a database.
  \-d THRESHOLD, \-\-threshold THRESHOLD
                        system threshold for ranking plot, default = 0.7
  \-c CONFIGFILENAME, \-\-config CONFIGFILENAME
                        use alternative config file
  \-l, \-\-license         show license
  \-s, \-\-settings        show settings only
  \-q, \-\-quiet           do not show settings
.ft P
.fi
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B Note:
.INDENT 7.0
.IP \(bu 2
that you can use several options at the same time. The sequence is of no importance.
.IP \(bu 2
each time in this documentation you read \(aqpython\(aq as part of a command then on a windows platform use \(aqpython.exe\(aq (without the quotes).
.IP \(bu 2
On OSX and linux you need not specify the python interpreter at all. bioplot.py can be run using: ./bioplot.py
.UNINDENT
.UNINDENT
.sp
As you can tell from the list of options you need to choose what type of graph you want.
Bioplot can produce several plots in a row with just one invocation of the program. E.g. if you run:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-Z \-A \-E \-i some_input_data_file.txt yet_another_input_file.txt
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
the program will produce a zoo plot, an accuracy plot and an EER plot one after the other.
.sp
The input data has to be in a specific format. Actually there are 2 types of format allowed. See \fI\%Data files\fP below.
All plots can be saved. This happens automagically as well, as soon as you click on a plot and press a key, but depending
on the screen size you get more useful results if this happens under your control.
.sp
If you do not provide an input file, the program uses input/testdata_A.txt.
This makes it is easy to try the program. You can try the multi experiment capabilities
of the program by choosing input/testdata_AB.txt or input/testdata_ABC.txt.
.sp
Bioplot when run reads its config file (by default called \(aqbioplot.cfg\(aq) and uses the settings it finds in there.
If no config file exists it uses a set of default values. Have a look at the config file. The program has quite a lot
of settings you may want set to your own liking.
.SS Auto store
.sp
The program does a bit more than its command line arguments suggest.
You will notice this when you run it. It will for instance store all the
target and non target scores it distills from the data file you pass
to it and write them in text files (unless you set saveScores to False in bioplot.cfg).
For this to be usefull always add a name for the experiment you were running to the command line,
otherwise the prefix \(aqtest\(aq is used.
.sp
The experiment name is used in the titles of the plots and in the filenames bioplot produces.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-Z \-e myFirstExperiment
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can use these file for further processing. The experiment name you specify is used as part
of the filename: <exp_name>_<meta_value>_non_target.txt, <exp_name>_<meta_value>_target.txt.
The <meta_value> is taken from the last column of data in the data files:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
myFirstExperiment_conditionA_target.txt
myFirstExperiment_conditionA_non_target.txt
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The files are stored in the directory specified by \(aqoutputPath\(aq in
bioplot.cfg in it\(aqs [cfg] section. The default path will be \(aqoutput\(aq in
the current directory. You can change this behaviour in bioplot.cfg via the following settings:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[cfg]
outputPath = output
saveScores = True
alwasySave = False
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If the output path does not exist, bioplot will try to create it. Note: this may require privileges to be set propperly.
If you run the program repeatedly using the same experiment name, the scores are not saved again, saving some processing
time, unless you add the following setting to the config file:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[cfg]
alwaysSave = True
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The default value is False as it is expected that you will run bioplot several times with the same data (hoping this
will speed things up a bit). Note, that if you do not change the experiment name but do change the data, if alwaysSave = False,
the data extracted will not reflect the analysis of the (new) data file used. If you want to have new versions of these files,
you need to delete them before running bioplot.py again (or set alwaysSave to True).
.sp
If you choose to plot a zoo plot, the labels which fall within the doves, chameleons, worms and phantom quartiles
are saved in individual text files: <exp_name>_chameleons.txt, <exp_name>_doves.txt, <exp_name>_phantoms.txt and
<exp_name>_worms.txt. This automatically documents all outliers.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
myFirstExperiment_chameleons.txt
myFirstExperiment_doves.txt
myFirstExperiment_phantoms.txt
myFirstExperiment_worms.txt
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The content of output/myFirstExperiment_chameleons.txt looks like this:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
# label, metavalue, average_target_score, average_non_target_score, nr_of_target_scores, nr_of_non_target_scores, average_target_score_stdev, average_non_target_score_stdev
116 conditionB \-0.563504928571 \-1.07438894059 14 202 \-0.502857 1.25531696972
1118 conditionA \-0.264748666667 \-1.07494067871 30 249 2.000000 \-0.733892414191
226 conditionB \-0.555475 \-1.06827121759 1 216 0.010000 \-0.717001903115
1066 conditionA \-0.399073347826 \-1.06514522689 23 238 1.985593 0.181852885922
1066 conditionB \-0.205634 \-1.07308166403 12 253 2.000000 0.234369371589
3146 conditionB \-0.542155 \-1.07573414493 2 207 \-0.588718 \-0.808330988919
1096 conditionB \-0.47514326087 \-1.06535841905 23 210 2.000000 \-0.964818041437
1096 conditionA \-0.3189693125 \-1.06838052917 32 240 2.000000 \-0.825256532013
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The labels with a normalized standard deviation for their target scores or their non target scores bigger than a maximum standard
deviation are stored in a file <exp name>_limited.txt together with the violating score (have a look at the
rst_zooplot page for a general understanding of how the plot is made).
.sp
Assuming these settings:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
[zoo]
limitStdDevs = True
maxStdDev = 2.0
minStdDev = 0.01
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
the following command will produce the file \(aqmyFirstExperiment_limited.txt\(aq where the standard deviations are limited to a maximum value of 2 * unit std dev:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-Z \-i input/testdata_AB.txt \-e myFirstExperiment
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This will result into the following content in output/myFirstExperiment_limited.txt (shown partially):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
# label, metavalue, average_target_score, average_non_target_score, nr_of_target_scores, nr_of_non_target_scores, average_target_score_stdev, average_non_target_score_stdev
223 conditionA \-0.6239485 \-1.0987091784 2 213 \-0.294772 2.0
1129 conditionB \-0.481959666667 \-1.09234021145 12 227 2.000000 \-0.487198065101
57 conditionB \-1.08974608333 \-1.07881886634 12 202 \-0.719556 \-2.0
609 conditionB \-0.614352277778 \-1.07410090099 18 202 \-0.482100 2.0
1130 conditionB \-0.0728461428571 \-1.11323731633 21 196 2.000000 0.0208211034315
223 conditionB \-0.59144 \-1.14699159184 1 196 0.010000 2.0
1112 conditionA \-0.416871 \-1.103248548 34 250 2.000000 1.17620375038
1116 conditionA \-0.4189825625 \-1.1624667598 32 204 1.920131 2.0
335 conditionA \-0.3849315 \-1.15242778855 2 227 \-0.545516 2.0
1123 conditionB \-0.323721045455 \-1.07929185124 22 242 2.000000 \-0.212778963187
1115 conditionA \-0.491315227273 \-1.14706272124 22 226 2.000000 1.42789578734
1131 conditionA \-0.3203315 \-1.09662913636 34 220 2.000000 0.0194105127029
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Note that the default setting for maxStdDev = 6.0 * unit std dev.
.SS Save plots
.sp
Any plot you produce will be saved to disk as a png\-file in the output directory as soon as you click on the plot
(to get the window focus) and press a key. The experiment name is used as part of the plot name. Note, it is important to maximise the plot\(aqs window to
get a proper layout of all elements in the plot! If you press \(aqs\(aq, you will be presented with a menu which
will allow you to save the plot anywhere you choose to. If you press a different key, the plot will be saved locally in
the directory specified by outputPath. This happens any time you press a key!
.sp
Note: the keys l, k, g, s and f are predefined keys of the gui. They provide additional functions.
With them you can:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
g: toggle grid on / off
k: toggle between linear horizontal scale and logaritmic horizontal scale
l: toggle between linear vertical scale and logaritmic vertical scale
s: open save menu
f: toggle between standard size and full screen
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Any key will make that the file is saved in its window\(aqs dimensions.
To get a nice plot it is wise to maximise the window and then press a key. Then close the window.
.sp
Note that using l and k may lead to warnings if e.g. the scores contain negative values (the log is only defined for numbers >= 0).
.SS Data files
.sp
The command line allows to specify a filename or a list of filenames and a type. The default type is \(aqtype3\(aq which corresponds to a text file
with 7 fields. You need not specify type3 as it is a default. This data type allows you to specify scores of
experiments where multiple files are used to make a model whether this is a training model or a test model.
Note, if you do not provide an input file, the program uses input/testdata_A.txt.
.sp
The data file should contain data in a format like this example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
803742 17133729a.wav 803593 16842970b.wav 2.108616847991943 FALSE META_VAL1
148407 47968376b.wav 898232 08087650a.wav 0.336018745422363 FALSE META_VAL3
179408 34192626a.wav 803721 16749939b.wav 1.263523664188385 FALSE META_VAL2
803442 48588750a.wav 803442 15560933b.wav 4.423274517059326 TRUE  META_VAL2
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Separation by comma\(aqs is also accepted.
This can be mixed as in:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
803742,17133729a.wav,803593,16842970b.wav,2.108616847991943,FALSE,META_VAL1
148407,47968376b.wav,898232,08087650a.wav,0.336018745422363,FALSE,META_VAL3
179408 34192626a.wav 803721 16749939b.wav 1.263523664188385 FALSE META_VAL2
803442,48588750a.wav,803442,15560933b.wav,4.423274517059326,TRUE, META_VAL2
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
field 1: string: label identifying a subject (training data).
.sp
field 2: string: name of data file containing biometric features or raw data originating from the subject denoted by field 1 used for making a test/evaluation model. In the example you see a wav\-file, but this can be any string identifying a file or feature set.
.sp
field 3: string: label identifying a subject (test/evaluation data).
.sp
field 4: string: name of data file containing biometric features or raw data originating from the subject denoted by field 3 used for training the reference model. In the example you see a wav\-file, but this can be any string identifying a file or feature set.
.sp
field 5: string: floating point value: score of trial.
.sp
field 6: boolean: ground truth.
.sp
field 7: meta data value for the experiment.
.sp
Field 7 can be used to contrast scores of experiments in most plots.
So if you have 2 experiments where you change one variable, when doing a cross
identification test, the meta value can be used to group the experiment\(aqs scores.
.sp
Let\(aqs have a look at an example data line:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
803742 17133729a.wav 803593 16842970b.wav 2.108616847991943 FALSE META_VAL1
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This line describes the result of a speaker verification experiment in which was used:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
\- as training audio samples from the file 17133729a.wav
\- the label 803742 to identify the speaker in the training audio
\- the label 803593 to identify the speaker in the test/evaluation/disputed utterance
\- as evaluation audio samples from the file 16842970b.wav
\- the tag FALSE to denote that the training speaker is not the test/eval speaker
\- META_VAL1 as a label to identify the conditions of the experiment (e.g. this is an experiment
  with training duration X seconds and test duration Y seconds, you may have various
  variables and use other meta labels to identify those).
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The result of the verification experiment was a score of 2.108616847991943.
.sp
If you have done experiments where the training model was made from several audio files, then
in field 4 use some symbolic name.
.sp
Note: bioplot.py uses the filenames in the data file only as symbols, the sample data in
the files is not used in any way.
.sp
E.g. you run an experiment with gender as the main variable and you collect scores of male
to male and female to female comparisons. You need to set the meta value for each score
accordingly. The meta value field allows bioplot to distinguish
between the two conditions and it will in essence plot 2 plots in one overview.
If the same label occurs on more than one occasion in the score list with different meta
values, then the points in a zoo plot with corresponding labels are interconnected
(see interconnectMetaValues setting in bioplot.cfg under [zoo]) . This makes it easy
to see what the effect of an individual label is when changing the experiment\(aqs condition.
.sp
Field 7 must be present. If you don\(aqt want to contrast experiments, then give all lines
the same meta value. Any string of characters (excluding white space) will do except
the special characters mentioned below under \(aqKnown Issues\(aq.
.sp
File type \(aqtype2\(aq is deprecated as of august 2015.
.sp
File \(aqtype1\(aq is a database based data format and meant as an example on how to use bioplot in combination with a
database of scores. Specify \(aqdatabase\(aq as filename on the command line.
.sp
Example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-i database \-t type1 \-e \(aqdata taken from db\(aq \-Z
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You will have to adapt the query in the function _readFromDatabase and the function
_decodeType1Results in data.py to your own needs.
.sp
There are several data files (of type3) which are meant as examples to play around
with: input/testdata_A.txt, input/testdata_B.txt, input/testdata_C.txt and input/testdata_ABC.txt
.sp
Example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
python bioplot.py \-e "condition A" \-i input/testdata_A.txt \-Z
python bioplot.py \-e "condition A and B" \-i input/testdata_A.txt input/testdata_B.txt \-Z
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
If you experience any difficulties reading your data file, we can either discuss this
via email or you can send the data it to me ( josbouten at gmail dot com ) so that I can have a look at it.
Please consider anonymizing the data before you send it to me by mail! Have a look at
rst_anonymize\&.
.sp
Note: bioplot.py does not allow as input data a text file and a database at the same time.
.SS Data exchange
.sp
bioplot formatted text files from linux, OSX and MS Windows platforms should be interchangeable without problems.
.SS Data anonymity
.sp
If you want to anonymize your data e.g. in case you want to exchange data file with somebody, have a look at rst_anonymize\&.
.SS Bugs and feature requests
.sp
If you have any questions or feature requests (no guarantees!) or find any bugs,
you can contact me at josbouten at gmail dot com.
.SH INSTALL
.SS Ms Windows
.sp
Follow the instructions found here: \fI\%http://matplotlib.org/users/installing.html\fP\&.
Download anaconda3\-4.2.0 (32 bit version) on W7 32 bits or the 64 bits version according to your machine\(aqs architecture.
Anaconda will install python 3.5 and a load of python modules amongst which numpy, matplotlib, pyplot and scikit\-learn. You will be able to run bioplot.py and do much more pythony things ;\-)
.sp
Download the zip file from the github page and unzip its contents to a local directory.
.sp
Copy bioplot.cfg_4_windows.txt to bioplot.cfg in the working directory (the directory where
bioplot.py is stored):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
copy bioplot.cfg_4_windows.txt bioplot.cfg
.ft P
.fi
.UNINDENT
.UNINDENT
.sp

changelog.txt
# V1.2.1 4 May 2017
Added some more instructions to install bioplot on OSX.
Thanks Jonas for the tips!

# v1.2 23 April 2017
Changed several copyright notices.
Added the number of target and non targets experiment counts to several plots.
This will make clear how many scores were used to plot them.
Added argument parser to anonymize.py

# v1.1 12 March 2017
Writing plot to file repaired.

# v1.0 10 March 2017
Added raw2bioplot.py conversion program.

# v0.9.9.2 29 Nov 2017
Converted tools to python3
Moved argparse handling to a method

# v0.9.9.1 14 Nov 2016
Resolved init super class bug in alexanderzoo.py

# v0.9.9 7 Nov 2016
Added writing plot to file
Converted all code to python3.5

# v0.9.8b 19 August 2016
Path to Users directory on OSX changed in doc.

# v0.9.8a 3 August 2016
Small text changes in the doc.

# v0.9.8 25 March 2016
Added support for multiple input files.

# v0.9.7 7 Febr 2016
Replaced optparse by argparse in bioplot.py,
bv2bioplot.py, sretools2bioplot and vocalise2bioplot.py
Introduced argument -L for labels in bioplot.py.

# v0.9.6.4 22 November 2015
bug: double labels in legend of some plots

# v0.9.6.3 22 November 2015
Added diagonal line to DET-plot, updates images in doc.
Method name bug in zoo.py

# v0.9.6.2 22 November 2015
Added some more small things to the doc.

# v0.9.6.1 22 November 2015
Added some small things to the doc.

# v0.9.6 12 November 2015
Added roc plot and det plot.
Changed config parameter names showMinCllrValues into showMinCllr
Changed config parameter names showCllrValues into showCllr

# v0.9.5 27 September 2015
Zoo object does not inherit cllr object anymore (there was no need to do so anyway).
Cleaned up code thus adhering to python coding standards a bit more.

# v0.9.4 Tue September 20. 2015
Improved positioning of labels in zoo plot, simplified code to do so quite a bit.
Improved position of legend in accuracy and ranking plot.
Improved scaling of reference and unit ellipses in zoo plot.
Added helper circles to zoo plot.
Paired colors of plot lines in eer plot and histograms.
Added cllr to legend of EER plot for multi and single condition experiments.
Added cllrMin to legend of Zoo and EER plot for multi and single condition experiments.
Added quite a bit of text to the doc(rst) files.
Added some conversion scripts: bv2bioplot.py, sretools2bioplot.py and vocalize2bioplot.py
Updated anonymize.py to use the sha256 function.
Made label click more selective. Added label fields.
Cleaned up settings files.
On OSX install of scikit-learn is required.
Changed html_theme to 'bootstrap' for sphinx html doc (see https://github.com/ryan-roemer/sphinx-bootstrap-theme).
Generated man pages for OSX and linux in bioplot/doc/_build/man/bioplot.1

# v0.9.3 August 2015
Added EER and Cllr to legend of Zoo plot for multi and single condition experiments.
Added UPDATE section to docs.
Added screen resolution variable to config file to properly position legend text.
Added choice for number of bins in histogram.
Added threading to writing target and non target data to disk.
Expanded use of meta data specific results in zoo plot.
Added nr of targets and nr of non targets to animal files.
Added nr of targets and nr of non targets to labels in zoo plot.
Added stdev scores to labels in zoo plot.
When adding labels to the command line, the connecting lines are fatter and less opaque then for the other labels.
Replaced use of color map by list of user defined colours (in hex or decimal).
Legend is placed in inner most quartile in zoo plot.
Reworked opacity calculations.
Removed type2 data support. Added some comments on how to reformat data to type3.
Repaired some minor bugs.

# v0.9.2 Fri May 8 2015
Added ability to choose a config file from the prompt.
Added flag for running Windows.

# v0.9.1 Fri April 24 2015
Scores written to <project>_target.txt and <project>_non_target.txt are now in one column in stead of
in sets.

# v0.9 Fri Oct 24 20:57:54 CEST 2014
Added cumulative score plot / EER plot and Tippett plot.
Added matrix plot.
Changed names of all plot routines.
Removed kernel function from histogram. There is only so much
a single gaussian can model.
Added rst and html manual making the manual.txt file superfluous.

# v0.8.1 Tue Oct  7 21:31:13 CEST 2014
For OSX changed alpha value of title and quartile animal labels.

# v0.8 Tue Oct  7 01:10:47 CEST 2014
Added flag for OSX.

# v0.7 Sun Oct  5 18:49:12 CEST 2014
Made preparations for storing at github.

# v0.6 Sept 29 2014
Moved a lot of functions common to boutenZoo and alexanderZoo to zoo.py
Added annotations and comments.

# v0.5 Sept 21 2014
Created zoo, format and data object to make things more OO.

# v0.4 Sept 12 2014
Added config file 'plot.cfg' containing default settings.

# v0.3, Sept 9 2014
Added click events. Clicking in zoo plot will show a list of nearest labels including
their average target and non target score. All labels within 1/10 of the minimum of the
height or width of an ellipse are shown.
Keys in other plots will save the file.
Added histograms (normal and cumulative) to zoo plot.

# v0.2, August 2014
Added 'agm and aim std deviations' to zoo plot after reading Anil Alexander's 
abstract for a paper he submitted to IAFPA 2014, Zurich, Switserland.

# v0.1, Sept 2013
Created code for ranking and accuracy plot.
Added zoo plot after reading a publication by Yager et al. about their 
variant on Doddington's zoo.


bioplot.py
#!/usr/bin/env python3

__author__ = 'drs. ing. Jos Bouten'

'''

    bioplot.py

    Main program demonstrating several plots that can be used
    when evaluating the performance of a biometric system.


    Copyright (C) 2014 Jos Bouten ( josbouten at gmail dot com )

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

'''
import argparse
import sys

from config import Config
from data import Data
from alexanderzoo import AlexanderZoo
from boutenzoo import BoutenZoo
from det import Det
from eer import Eer
from tippett import Tippett
from histogram import Histogram
from accuracy import Accuracy
from ranking import Ranking
from roc import Roc
from matrix import MatrixPlot
from version import Version
from utils import sanitize
from license import License
from os.path import basename


def printConfig(theseOptions, thisConfig):
    if thisConfig.getFileNotFound():
        print(("Config info taken from 'default values':\n%s" % thisConfig.toString()))
    else:
        if theseOptions.configFilename:
            print(("Config info as read from '%s':\n%s" % (theseOptions.configFilename, thisConfig.toString())))
        else:
            print(("Config info taken from 'default values':\n%s" % thisConfig.toString()))


def parseArguments():
    v = Version()
    progName = basename(sys.argv[0])
    version = "This is bioplot.py version %s, Copyright (C) 2014: Jos Bouten" % v.getVersion()
    parser = argparse.ArgumentParser(description="%s [plot type] [<label1> <label2> <label3> ...]\n\
    bioplot.py version %s.\n\
    This program comes with ABSOLUTELY NO WARRANTY; for details run `bioplot.py -l\'.\n\
    This is free software, and you are welcome to redistribute it\n\
    under certain conditions; type `bioplot.py -l\' for details.\n\
    This program was written by Jos Bouten.\n\
    You can contact me via josbouten at gmail dot com." % (progName, version))
    parser.add_argument('-Z', '--zoo', action="store_true", dest="plotZoo", help="show zoo plot")
    parser.add_argument('-A', '--accuracy', action="store_true", dest="plotAccuracy", help="show accuracy plot")
    parser.add_argument('-D', '--det', action="store_true", dest="plotDet", help="show Det plot")
    parser.add_argument('-E', '--eer', action="store_true", dest="plotEer", help="show EER plot")
    parser.add_argument('-T', '--tippet', action="store_true", dest="plotTippet", help="show Tippett plot")
    parser.add_argument('-M', '--matrix', action="store_true", dest="plotMatrix", help="show matrix plot")
    parser.add_argument('-O', '--roc', action="store_true", dest="plotRoc", help="show roc plot")
    parser.add_argument('-R', '--ranking', action="store_true", dest="plotRanking", help="show ranking plot")
    parser.add_argument('-C', '--histogramc', action="store_true", dest="plotHistCum", help="show cumulative histogram")
    parser.add_argument('-H', '--histogram', action="store_true", dest="plotHist", help="show histogram")
    parser.add_argument('-k', '--kernel', action="store_true", dest="plotKernel",
                        help="show kernel estimate in histogram")
    parser.add_argument('-L', '--label', dest='labels', type=str, nargs='+', help="add labels to plot")
    parser.add_argument('-e', '--exp', action="store", dest="expName", default='test',
                        help="name of experiment used in plot title, default = test")
    parser.add_argument('-i', '--inputfile', action="store", dest="filenames", nargs='+',
                        default=['input/testdata_A.txt'],
                        help="filename of filenames of data file(s) or name of database, default = input/testdata_A.txt")
    parser.add_argument('-t', '--type', action="store", dest="dataType", default='type3',
                        help="type of data, default = type3, use 'database' if you want to read data from a database.")
    parser.add_argument('-d', '--threshold', action="store", dest="threshold", type=float, default=0.7,
                        help="system threshold for ranking plot, default = 0.7")
    parser.add_argument('-c', '--config', action="store", dest="configFilename", default='bioplot.cfg',
                        help="use alternative config file")
    parser.add_argument('-l', '--license', action="store_true", dest="showLicense", help="show license")
    parser.add_argument('-s', '--settings', action="store_true", dest="showOptions", help="show settings only")
    parser.add_argument('-q', '--quiet', action="store_true", dest="quiet", help="do not show settings")
    return parser.parse_args()


#
#  Main
#

# Define command line parser and get cli arguments.
args = parseArguments()

# Let's handle any request for the license first.
# We stop the program after that.
if args.showLicense:
    l = License('LICENSE.txt')
    l.showLicense()
    exit(0)

# Name of the experiment, used as _title in plots.
expName = args.expName

# We do not like spaces!
filenames = sanitize(args.filenames)
# filename = args.filename
dataType = args.dataType

# Threshold used by biometric system to make a decision
# only of interest if you want to plot the systems Accuracy.
threshold = args.threshold

config = Config(args.configFilename)

debug = config.getDebug()

if args.quiet:
    config.setShowConfigInfo(False)

if args.showOptions:
    print("Ignoring all command line parameters except -s")
    printConfig(args, config)
    exit(1)

if config.getShowConfigInfo():
    printConfig(args, config)

print(("Reading data from:",))
for el in args.filenames:
    print((el,))
print()

data = Data(config, expName, threshold, dataType, debug, filenames)

if config.getSaveScores():
    # Write data to text files sorted by meta values.
    data.writeScores2file(data.getTargetScores(), expName, '_target.txt')
    data.writeScores2file(data.getNonTargetScores(), expName, '_non_target.txt')

if args.labels:
    if len(args.labels) > 0:
        data.setLabelsToShowAlways(args.labels)

if args.plotAccuracy:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        accuracy = Accuracy(data, config, expName, debug)
        accuracy.plot()
    else:
        print("Not enough data.")

if args.plotDet:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        det = Det(data, config, expName, debug)
        det.plot()
    else:
        print("Not enough data.")

if args.plotEer:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        eer = Eer(data, config, expName, debug)
        eer.plot()
    else:
        print("Not enough data.")

if args.plotHistCum:
    # Interested in EER plot? Then plot a cumulative histogram of the scores.
    # More crude than eer.plot and not differentiating between meta values.
    histogram = Histogram(data, config, expName, 'cumulative', debug)
    histogram.plot()

if args.plotHist:
    # Show histogram for data split by meta data value.
    useMeta = True
    if args.plotKernel:
        # Add all target and non target data together, i.e. do not use meta data label info.
        useMeta = False
    histogram = Histogram(data, config, expName, 'normal', debug, useMeta)
    histogram.plot()

if args.plotMatrix:
    matrix = MatrixPlot(data, config, expName, debug)
    matrix.plot()

if args.plotRanking:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        ranking = Ranking(data, config, expName, debug)
        ranking.plot()
    else:
        print("Not enough data.")

if args.plotRoc:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        roc = Roc(data, config, expName, debug)
        roc.plot()
    else:
        print("Not enough data.")

if args.plotTippet:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        tippet = Tippett(data, config, expName, debug)
        tippet.plot()
    else:
        print("Not enough data.")

if args.plotZoo:
    if (len(data.getTargetCnt()) > 0) and (len(data.getNonTargetCnt()) > 0):
        if config.getBoutenStyle() is True:
            zoo = BoutenZoo(data, config, expName, debug)
            zoo.plot()
        else:
            zoo = AlexanderZoo(data, config, expName, debug)
            zoo.plot()
    else:
        print("Not enough data.")
